core/src/main/java/org/infinispan/commands/remote/ClusteredGetAllCommand.java

```diff
@@ -0,0 +1,190 @@
+ *
+ * @author Radim Vansa &lt;rvansa@redhat.com&gt;
+ */
+public class ClusteredGetAllCommand<K, V> extends LocalFlagAffectedRpcCommand {
+   public static final byte COMMAND_ID = 46;
+   private static final Log log = LogFactory.getLog(ClusteredGetAllCommand.class);
+   private static final boolean trace = log.isTraceEnabled();
+
+   private List<?> keys;
+   private GlobalTransaction gtx;
+
+   private InvocationContextFactory icf;
+   private CommandsFactory commandsFactory;
+   private InterceptorChain invoker;
+   private TransactionTable txTable;
+   private InternalEntryFactory entryFactory;
+   private Equivalence<? super K> keyEquivalence;
+
+   ClusteredGetAllCommand() {
+      super(null, null);
+   }
+
+   public ClusteredGetAllCommand(String cacheName) {
+      super(cacheName, null);
+   }
+
+   public ClusteredGetAllCommand(String cacheName, List<?> keys, Set<Flag> flags,
+         GlobalTransaction gtx, Equivalence<? super K> keyEquivalence) {
+      super(cacheName, flags);
+      this.keys = keys;
+      this.gtx = gtx;
+      this.keyEquivalence = keyEquivalence;
+   }
+
+   public void init(InvocationContextFactory icf, CommandsFactory commandsFactory,
+         InternalEntryFactory entryFactory, InterceptorChain interceptorChain,
+         TransactionTable txTable, Equivalence<? super K> keyEquivalence) {
+      this.icf = icf;
+      this.commandsFactory = commandsFactory;
+      this.invoker = interceptorChain;
+      this.txTable = txTable;
+      this.entryFactory = entryFactory;
+      this.keyEquivalence = keyEquivalence;
+   }
+
+   @SuppressWarnings("unchecked")
+   @Override
+   public Object perform(InvocationContext ctx) throws Throwable {
+      acquireLocksIfNeeded();
+      // make sure the get command doesn't perform a remote call
+      // as our caller is already calling the ClusteredGetCommand on all the relevant nodes
+      GetAllCommand command = commandsFactory.buildGetAllCommand(keys, flags, true);
+      InvocationContext invocationContext = icf.createRemoteInvocationContextForCommand(command, getOrigin());
+      Map<K, CacheEntry<K, V>> map = (Map<K, CacheEntry<K, V>>) invoker.invoke(invocationContext, command);
+      if (trace) log.trace("Found: " + map);
+
+      if (map == null) {
+         return null;
+      }
+
+      //this might happen if the value was fetched from a cache loader
+      List<InternalCacheValue<V>> values = new ArrayList<>(keys.size());
+      for (Object key : keys) {
+         if (map.containsKey(key)) {
+            CacheEntry<K, V> entry = map.get(key);
+            InternalCacheValue<V> value;
+            if (entry instanceof InternalCacheEntry) {
+               value = ((InternalCacheEntry<K, V>) entry).toInternalCacheValue();
+            } else if (entry != null) {
+               value = entryFactory.createValue(entry);
+            } else {
+               value = new ImmortalCacheValue(null);
```

Comment:
Also I am sorry but I mixed up the usage of null.  If a CacheValue with a null value is sent back that is treated as if the entry was not present in the cache when this node owned the key.  If there is no entry returned for a given key that is when it assumes this node lost ownership.